src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/prisonersearch/PrisonerSearchResultsProcessor.kt

```diff
@@ -0,0 +1,104 @@
+import uk.gov.justice.digital.hmpps.prisonersearch.model.PrisonerMatches
+import uk.gov.justice.digital.hmpps.sendlegalmailtoprisonsapi.prisonersearch.PrisonerSearchDataSource.GLOBAL_SEARCH
+import uk.gov.justice.digital.hmpps.sendlegalmailtoprisonsapi.prisonersearch.PrisonerSearchDataSource.MATCH_PRISONERS
+import uk.gov.justice.digital.hmpps.sendlegalmailtoprisonsapi.prisonersearch.PrisonerSearchType.NAME_DOB
+import uk.gov.justice.digital.hmpps.sendlegalmailtoprisonsapi.prisonersearch.PrisonerSearchType.PRISON_NUMBER
+
+@Component
+class PrisonerSearchResultsProcessor(
+  private val telemetryClient: TelemetryClient,
+  private val prisonerSearchResultsLogger: PrisonerSearchResultsLogger,
+) {
+
+  fun processSearchResults(
+    prisonerMatches: PrisonerMatches,
+    prisonerSearchRequest: PrisonerSearchRequest,
+    caseload: String
+  ) {
+    prisonerSearchResultsLogger.logResults(prisonerMatches, prisonerSearchRequest, caseload)
+
+    val prisoners = prisonerMatches.matches.map { it.prisoner }
+    val prisonerMatchCount = PrisonerMatchCount.of(prisoners, prisonerSearchRequest)
+    val customDimensions = PrisonerSearchCustomDimensions(
+      dataSource = MATCH_PRISONERS,
+      searchType = prisonerSearchRequest.searchType,
+      numberOfResults = prisoners.size,
+      exactMatchCount = prisonerMatchCount.mainDetailsCount,
+      aliasExactMatchCount = prisonerMatchCount.aliasDetailsCount
+    ).asMap()
+
+    telemetryClient.trackEvent("prisoner-search", customDimensions, null)
+  }
+
+  fun processSearchResults(
+    pagePrisoner: PagePrisoner,
+    prisonerSearchRequest: PrisonerSearchRequest,
+    caseload: String
+  ) {
+    prisonerSearchResultsLogger.logResults(pagePrisoner, prisonerSearchRequest, caseload)
+
+    val prisoners = pagePrisoner.content ?: emptyList()
+    val prisonerMatchCount = PrisonerMatchCount.of(prisoners, prisonerSearchRequest)
+    val customDimensions = PrisonerSearchCustomDimensions(
+      dataSource = GLOBAL_SEARCH,
+      searchType = prisonerSearchRequest.searchType,
+      numberOfResults = prisoners.size,
+      exactMatchCount = prisonerMatchCount.mainDetailsCount,
+      aliasExactMatchCount = prisonerMatchCount.aliasDetailsCount
+    ).asMap()
+
+    telemetryClient.trackEvent("prisoner-search", customDimensions, null)
+  }
+}
+
+private val PrisonerSearchRequest.searchType: PrisonerSearchType
+  get() = if (prisonNumber != null) PRISON_NUMBER else NAME_DOB
+
+data class PrisonerMatchCount(
+  val mainDetailsCount: Int,
+  val aliasDetailsCount: Int
+) {
+  companion object {
+    fun of(prisoners: List<Prisoner>, prisonerSearchRequest: PrisonerSearchRequest): PrisonerMatchCount {
+      val totalPrisoners = prisoners.size
+
+      if (prisonerSearchRequest.searchType == PRISON_NUMBER) {
+        // Count the prisoners with a match on prison number & main name, or prison number & alias name
+        val prisonersNotMatchingMainDetails = prisoners.filterNot {
```

Comment:
We need to make sure we dont double count - ie. a prisoner record where the prison number matches and the main name matches; plus that same prisoner has an alias with the same name - in that case we dont want to count a match on the main details AND a match on the alias - hence the filtering approach